api/sip/client.py

```diff
@@ -705,14 +707,22 @@ def read_message(self, max_size=1024*1024):
         data = ""
         tmp = ""
         while not done:
-            tmp = self.socket.recv(4096)
+            try:
+                tmp = self.socket.recv(4096)
+            except socket.timeout:
+                raise socket.error
+            except socket.error, e:
+                raise IOError("Socket error occurred receiving SIP response: %s" % e)
+            except Exception, e:
+                raise IOError("Some other error occurred with SIP response: %s" % e.message)
             data = data + tmp
-            if not tmp:
+            if not data:
                 raise IOError("No data read from socket.")
             if ord(data[-1]) == 13 or ord(data[-1]) == 10:
                 done = True
             if len(data) > max_size:
                 raise IOError("SIP2 response too large.")
+            # Do we need to test for a malformed response with no proper line termination?
```

Comment:
I don't know how we'd distinguish between a response with no line termination and an incomplete response that times out.

Reply:
I left the comment in from my working/testing copy, because it was part of my trying to figure out what was going on--and when using if not data: there is a possibility of an infinite loop. So maybe the best reason for leaving the condition at if not tmp as you suggested. tmp being empty would be caught, indicating the connection was closed remotely. Would a socket timeout _with data having content_ (nothing else received after receiving something) indicate a malformed/improperly terminated response? 

This may be more academic than it deserves treatment. But it was in my mind, so I left the comment.

Reply:
I'm not 100% on this, but... we're asking for four kilobytes of data. I don't think there's a maximum size of a SIP message (although we give up after receiving a megabyte) but I've never seen one longer than maybe one kilobyte.

`man recv` says "The receive calls normally return any data available, up to the requested amount,  rather  than  waiting  for receipt of the full amount requested."

If we ask for four kilobytes of data, and we get less than that, it's pretty likely that's the whole message. But I _think_ if we ask again we're going to read zero bytes rather than getting a timeout. At that point `tmp` will be empty (even though `data` is non-empty) and we'll proceed as normal.